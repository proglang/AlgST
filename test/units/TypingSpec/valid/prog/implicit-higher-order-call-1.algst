calc : ({Int} -> Int) -> Int

check : Int
check = calc (implicit[Int])

synth : Int
synth =
  let r = calc (implicit[Int]) in
  r
