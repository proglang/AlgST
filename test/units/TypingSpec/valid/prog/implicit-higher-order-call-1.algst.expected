calc#M.2
|
`- SignatureDecl
   |
   `- type
      |
      `- Type.Arrow U
         |
         +- Type.Arrow ? U
         |  |
         |  +- Type.Type
         |  |  |
         |  |  `- TypeRef
         |  |     |
         |  |     `- Int
         |  |
         |  `- Type.Type
         |     |
         |     `- TypeRef
         |        |
         |        `- Int
         |
         `- Type.Type
            |
            `- TypeRef
               |
               `- Int

check#M.0
|
`- ValueDecl
   |
   +- type
   |  |
   |  `- Type.Type
   |     |
   |     `- TypeRef
   |        |
   |        `- Int
   |
   +- params
   |
   `- definition
      |
      `- Exp.App
         |
         +- Exp.Var calc#M.2
         |
         `- Exp.Abs U «implicit-arg»#M.4
            |
            +- Type.Type
            |  |
            |  `- TypeRef
            |     |
            |     `- Int
            |
            `- Exp.App
               |
               +- Exp.TypeApp
               |  |
               |  +- Exp.Var implicit#Builtin.32
               |  |
               |  `- Type.Type
               |     |
               |     `- TypeRef
               |        |
               |        `- Int
               |
               `- Exp.Var «implicit-arg»#M.4

synth#M.1
|
`- ValueDecl
   |
   +- type
   |  |
   |  `- Type.Type
   |     |
   |     `- TypeRef
   |        |
   |        `- Int
   |
   +- params
   |
   `- definition
      |
      `- Exp.Exp
         |
         `- TcExpX.ValueCase
            |
            +- Exp.App
            |  |
            |  +- Exp.Var calc#M.2
            |  |
            |  `- Exp.Abs U «implicit-arg»#M.5
            |     |
            |     +- Type.Type
            |     |  |
            |     |  `- TypeRef
            |     |     |
            |     |     `- Int
            |     |
            |     `- Exp.App
            |        |
            |        +- Exp.TypeApp
            |        |  |
            |        |  +- Exp.Var implicit#Builtin.32
            |        |  |
            |        |  `- Type.Type
            |        |     |
            |        |     `- TypeRef
            |        |        |
            |        |        `- Int
            |        |
            |        `- Exp.Var «implicit-arg»#M.5
            |
            `- r#M.3
               |
               `- Exp.Var r#M.3

